aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,673 @@
+    /// [`HpkeKeyState::Expired`]. It is at operator discretion as to how long this should be.
+    pub active_duration: Duration,
+
+    /// How long the key remains in [`HpkeKeyState::Expired`] before being deleted. This should
+    /// be greater than the clients' HPKE key cache maximum age.
+    pub expired_duration: Duration,
+
+    /// The ciphersuite of the key.
+    pub ciphersuite: HpkeCiphersuite,
+}
+
+impl<C: Clock> KeyRotator<C> {
+    pub fn new(datastore: Arc<Datastore<C>>, hpke: Vec<HpkeKeyRotatorConfig>) -> Self {
+        Self { datastore, hpke }
+    }
+
+    /// Runs the key rotator.
+    ///
+    /// # Errors
+    ///
+    /// Errors on general datastore errors.
+    ///
+    /// This is not permissive of [`HpkeConfigId`] space exhaustion, i.e. if there are more rows
+    /// than what fit into a `u8`, this process will error. To avoid this, keep the list of managed
+    /// ciphersuites small, and the expiration duration less than the key age.
+    #[tracing::instrument]
+    pub async fn run(&self) -> Result<(), Error> {
+        self.datastore
+            .run_tx("global_hpke_key_rotator", |tx| {
+                let configs = self.hpke.clone();
+                Box::pin(async move { Self::run_hpke(tx, configs).await })
+            })
+            .await
+            .map_err(|err| err.into())
+    }
+
+    #[tracing::instrument(skip(tx))]
+    async fn run_hpke(
+        tx: &Transaction<'_, C>,
+        configs: Vec<HpkeKeyRotatorConfig>,
+    ) -> Result<(), DatastoreError> {
+        // Take an ExclusiveLock on the table. This ensures that only one key rotator
+        // replica writes to the table at a time and that each replica gets a consistent
+        // view of the table state.
+        //
+        // ExclusiveLock does not conflict with the AccessShare lock taken by table
+        // reads, i.e. other aggregator replicas can continue to refresh their key
+        // caches without being blocked.
+        tx.lock_global_hpke_keypairs().await?;
+
+        let current_keypairs = tx.get_global_hpke_keypairs().await?;
+        debug!(?current_keypairs, "before");
+        let mut available_ids = Self::available_ids(&current_keypairs);
+        for config in &configs {
+            Self::run_hpke_for_config(tx, &current_keypairs, &mut available_ids, config).await?;
+        }
+
+        // Defensive assertion: Check our transaction snapshot for at least one active keypair in
+        // in the table. If one is absent, committing the transaction would leave Janus unstartable
+        // so we should rollback.
+        let current_keypairs = tx.get_global_hpke_keypairs().await?;
+        debug!(?current_keypairs, "after");
+        if !current_keypairs
+            .iter()
+            .any(|keypair| keypair.state() == &HpkeKeyState::Active)
+        {
+            Err(DatastoreError::User(
+                anyhow!("unexpected state: no keypairs are active").into(),
```

Comment:
nit: This could fire if the database is empty and the `hpke` list in the config is empty. We could check if there are no algorithms chosen first, log a warning, and exit early to provide a nicer message in that case.

Reply:
My (eventual) intent is to provide a sane default for `hpke`, such that the `hpke` list is never empty.

Reply:
On further thought I don't think it's wise to provide a default.

If we provide a default, if the operator wants to introduce a non-default key, they must provide the default key in their `hpke:` stanza, otherwise the key rotator will forget about the default key.

I think instead I'll just make providing at least one ciphersuite mandatory. (The goal of #2147 is to make global keys a requirement, so this approach makes sense to me).